### DISCUSSION TITLE
Trying to learn how programming languages are put together and V is my case study

### DISCUSSION THREAD

**Discussion by iambenkay** [vlang/v#7344] [General]
TLDR; please help point me to a file in the codebase that will give me a hint as to how the PL is put together. Basically how the compiler is organised together.

So I saw V for the first time many months ago and the fact that it's syntax is a blend of my two favourite languages, Golang and Rust is just captivating. I would always swear I'd come back to make a PR but work has been really intense. Well I am ready to take that bold step in contributing to V and the first thing on my list is understanding how PLs are put together entirely. V is my case study particularly because of its fast build. Would love to know what makes it tick. Can anyone here point me to the location of the codebase where I can kick off my dreams?

**Comment by medvednikov** [comment]
You are in the right place, V's code is probably the easiest to understand.

This is a good start:

https://github.com/vlang/v/blob/master/CONTRIBUTING.md

(Code Structure section)

**Reply by iambenkay** [reply]
thank you so much!

**Reply by nova-nowiz** [reply]
Hey, I learned a bunch with this document and also with [this](https://softwareengineering.stackexchange.com/questions/165543/how-to-write-a-very-basic-compiler#comment318391_165543)  stackoverflow question so thank you!
But now I have a lot of questions xD

First of all, what is the most difficult part of creating a language @medvednikov? Is it parsing, type checking or generating?

Parsing seems really difficult to learn, there is a lot of terminology and it seems for now that I'm only scratching the surface. I know about PEGs and CFGs, know the major difference but don't know what's best as CFGs lioe EBNF are very popular but PEG is used by zig (a very recent and safe language, so I guess it has advantages).

For type chwcking, I know about typing rules but not at all about type theory and it seems quite complex tbh.

Generating seems difficult but maybe feasible with the use of LLVM IR.

As I'm a bit lost, I would like to know if possible the tools (both theoretical and pratical) that you used for v @medvednikov (I hope that this is not too big of a task, sorry in advance if it is :/)

**Reply by spytheman** [reply]
Imho, https://craftinginterpreters.com/ is pretty good resource if you are novice.

**Reply by div72** [reply]
> but PEG is used by zig

@Narice Not really experienced with parsing but python also uses PEG now.
https://www.python.org/dev/peps/pep-0617/

**Comment by JalonSolov** [comment]
https://interpreterbook.com/
http://scheme2006.cs.uchicago.edu/11-ghuloum.pdf
